include/xtensor/xeval.hpp

```diff
@@ -47,6 +47,65 @@ namespace xt
     {
         return std::forward<T>(t);
     }
+
+    /**************
+     * as_strided *
+     **************/
+
+    /**
+     * Force evaluation of xexpression and convert to the required layout.
+     * @return xarray or xtensor depending on shape type
+     *
+     * \code{.cpp}
+     * xarray<double, layout_type::row_major> a = {1,2,3,4};
+     * auto&& b = xt::as_strided(a); // b is a reference to a, no copy!
+     * auto&& c = xt::as_strided<layout_type::column_major>(a); // b is xarray<double> with the required layout
+     * auto&& a_cast = xt::cast<int>(a); // a_cast is an xexpression
+     * auto&& d = xt::as_strided(a_cast); // d is xarray<int>, not an xexpression
+     * auto&& e = xt::as_strided<layout_type::column_major>(a_cast); // d is xarray<int> with the required layout
+     * \endcode
+     */
+    template <layout_type L = layout_type::any, class E, class I = std::decay_t<E>>
+    inline auto as_strided(E&& e)
+        -> std::enable_if_t<has_data_interface<I>::value
+                            && ((I::static_layout == L)
+                                || (L == layout_type::any)),
+                            E&&>
+    {
+        return std::forward<E>(e);
+    }
+
+    namespace detail
+    {
+        constexpr layout_type layout_remove_any(const layout_type layout)
+        {
+            return layout == layout_type::any ? XTENSOR_DEFAULT_LAYOUT : layout;
+        }
+    }
+
+    /// @cond DOXYGEN_INCLUDE_SFINAE
+    template <layout_type L = layout_type::any, class E, class I = std::decay_t<E>>
+    inline auto as_strided(E&& e)
+        -> std::enable_if_t<((!has_data_interface<I>::value)
+                             || ((I::static_layout != L) && (L != layout_type::any)))
+                            && detail::is_array<typename I::shape_type>::value,
+                            xtensor<typename I::value_type,
+                                    std::tuple_size<typename I::shape_type>::value,
+                                    detail::layout_remove_any(L)>>
```

Comment:
I think these complicated conditions deserve more expressive metafunctions or `constexpr` functions, especially since  they're used in many places.

Reply:
Agree, I'll do that